### PR TITLE
Handle edge cases when deserializing error JSON

### LINKED ISSUE #239: [Bug]: IllegalStateException when deseralizing "error"
### Software Version

6.3.0

### Language Version

11

### Operating System

Linux

### What happened?

We have noticed a bug in the software that occurs every time there is a validation errors when performing a One-Call Buy Shipment with Purolator and they return multiple validation errors. 
The bug is related to the deserialization of the JSON errors, which is not functioning as expected.
Specifically, we have identified that when your code is deserializing the error node, it expects the message to be either a string or an array, but in this case, it is an object, it cashes in your internal code, and we do not get an EasyPostException.
This issue is causing us significant trouble as we are currently unable to move forward with our production process.

I have included the JSON that is causing the problem below, as an example:
```
{
    "error": {
        "code": "UNPROCESSABLE_ENTITY",
        "message": {
            "errors": [
                "1100156: From On Label address is an invalid Service Point.",
                "1100674: Postal Code XXX is only valid for XXX."
            ]
        },
        "errors": [
            {
                "message": "1100156: From On Label address is an invalid Service Point."
            },
            {
                "message": "1100674: Postal Code XXX is only valid for XXX."
            }
        ]
    }
}
```

Code of your Error class
package com.easypost.model;

```
import java.util.List;

public final class Error {
    private String message;
    private String code;
    private List<Error> errors;
    private String suggestion;
    private String field;

    public Error() {
    }

    void setMessage(String message) {
        this.message = message;
    }

    void setCode(String code) {
        this.code = code;
    }

    public String getMessage() {
        return this.message;
    }

    public String getCode() {
        return this.code;
    }

    public List<Error> getErrors() {
        return this.errors;
    }

    public String getSuggestion() {
        return this.suggestion;
    }

    public String getField() {
        return this.field;
    }
}
```

When your deserializer, do the work, the message node is not a string, neither an array but and object and it crash.
```
public final class ErrorDeserializer implements JsonDeserializer<Error> {
    public ErrorDeserializer() {
    }

    public Error deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        JsonObject jo = json.getAsJsonObject();
        JsonElement results = jo.get("error");
        Gson gson = new Gson();
        if (results == null) {
            Error error = new Error();
            error.setMessage("API did not return error details.");
            error.setCode("NO RESPONSE CODE");
            return error;
        } else {
            JsonElement errorMessage = results.getAsJsonObject().get("message");
            if (errorMessage.isJsonArray()) { // This does not handle cases where it is an object
                JsonArray jsonArray = errorMessage.getAsJsonArray();
                ArrayList<String> messages = new ArrayList();

                for(int i = 0; i < jsonArray.size(); ++i) {
                    messages.add(jsonArray.get(i).getAsString());
                }

                JsonPrimitive value = new JsonPrimitive(String.join(", ", messages));
                results.getAsJsonObject().add("message", value);
            }

            return (Error)gson.fromJson(results, Error.class);
        }
    }
}
```

### What was expected?

The library to return a EasyPostException and beeing able to parse Errors correctly.

### Sample Code

_No response_

### Relevant logs

```sh
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected STRING but was BEGIN_OBJECT at path $.message
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:226)
        at com.google.gson.Gson.fromJson(Gson.java:932)
        at com.google.gson.Gson.fromJson(Gson.java:1003)
        at com.google.gson.Gson.fromJson(Gson.java:975)
        at com.easypost.model.ErrorDeserializer.deserialize(ErrorDeserializer.java:53)
        at com.easypost.model.ErrorDeserializer.deserialize(ErrorDeserializer.java:16)
        at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:69)
        at com.google.gson.Gson.fromJson(Gson.java:932)
        at com.google.gson.Gson.fromJson(Gson.java:897)
        at com.google.gson.Gson.fromJson(Gson.java:846)
        at com.google.gson.Gson.fromJson(Gson.java:817)
        at com.easypost.http.Requestor.handleAPIError(Requestor.java:530)
        at com.easypost.http.Requestor.httpRequest(Requestor.java:514)
        at com.easypost.http.Requestor.request(Requestor.java:438)
        at com.easypost.service.ShipmentService.create(ShipmentService.java:58)
        at com.easypost.service.ShipmentService.create(ShipmentService.java:40)
        at com.easypost.service.ShipmentService$create.call(Unknown Source)
```

### PR DESCRIPTION
**Pull request by jchen293**
# Description

Handle edge cases when deserializing error JSON, close #239 

# Testing

Add two unit tests and pass

# Pull Request Type

Please select the option(s) that are relevant to this PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement (fixing a typo, updating readme, renaming a variable name, etc)

### CODE REVIEW

File: src/main/java/com/easypost/model/ErrorDeserializer.java [hunk 1]
```diff
@@ -12,8 +12,32 @@
...
 
 public final class ErrorDeserializer implements JsonDeserializer<Error> {
+    /**
+     * Recursively traverse an error JSON element and its sub-element(s), and extracts all
+     * error string values found into the specified string list.
+     *
+     * @param element the JSON element to traverse
+     * @param messagesList the list of strings to append found values to
+     */
+    private void traverseJsonElement(JsonElement element, ArrayList<String> messagesList) {
```
Comment (nwithan8) [inline src/main/java/com/easypost/model/ErrorDeserializer.java:25]: This works because Java is "pass by reference", so the `messageList` variable getting updated by each recursive call does reflect on the variable on line 65 below.

However, this won't work in other languages that don't pass by reference. Instead, can we make this recursive function actually return the updated/appended `messageList` variable so we don't have to worry about its reference?

**Review by nwithan8 — CHANGES_REQUESTED** [review]
See comment above.

File: src/main/java/com/easypost/model/ErrorDeserializer.java (hunk 1 above)
Comment (jchen293) [reply src/main/java/com/easypost/model/ErrorDeserializer.java:25]: This will work with the other six languages as they support `pass by reference` too or has `pass by reference` already for passing in an object into a method, some languages might require you to add an additional symbol, like in Go you would need to use a pointer, and in PHP you would need to use `&`.

**Review by nwithan8 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,9 @@
 # CHANGELOG
 
+## v6.4.0 Next Release
```
Comment (Justintime50) [inline CHANGELOG.md:3]: Nit: version numbers are picked out at release time based on what's included, this may not be the only thing changed, for example this is really a patch and not a feature so it will most likely be 6.3.1. Let's not pre-label versions in the future

File: src/main/java/com/easypost/model/ErrorDeserializer.java [hunk 3]
```diff
@@ -12,8 +12,32 @@
...
 import java.util.ArrayList;
+import java.util.Map.Entry;
 
 public final class ErrorDeserializer implements JsonDeserializer<Error> {
+    /**
+     * Recursively traverse an error JSON element and its sub-element(s), and extracts all
+     * error string values found into the specified string list.
+     *
+     * @param element the JSON element to traverse
+     * @param messagesList the list of strings to append found values to
```
Comment (Justintime50) [inline src/main/java/com/easypost/model/ErrorDeserializer.java:23]: question: this description doesn't make sense, the last 3 words seem to not belong here or are unfinished?

File: src/main/java/com/easypost/model/ErrorDeserializer.java [hunk 4]
```diff
@@ -36,18 +60,18 @@ public Error deserialize(final JsonElement json, final Type typeOfT,
...
-            JsonArray jsonArray = errorMessage.getAsJsonArray();
-            ArrayList<String> messages = new ArrayList<>();
-
-            for (int i = 0; i < jsonArray.size(); i++) {
-                messages.add(jsonArray.get(i).getAsString());
-            }
 
+        try {
+            ArrayList<String> messages = new ArrayList<>();
+            JsonElement errorMessageJson = results.getAsJsonObject().get("message");
```
Comment (Justintime50) [inline src/main/java/com/easypost/model/ErrorDeserializer.java:66]: question: how does this work if it's a list? We are trying to get it as an object so will that mean we no longer capture when something is a list?

Comment (nwithan8) [reply src/main/java/com/easypost/model/ErrorDeserializer.java:66]: It will take the value of the "message" key and turn it into a `JsonObject`, the root of any JSON element for the GSON library. In the `traverseJsonElement` function, it will handle if it's a list/array.

File: src/main/java/com/easypost/model/ErrorDeserializer.java (hunk 3 above)
Comment (jchen293) [reply src/main/java/com/easypost/model/ErrorDeserializer.java:23]: `specified string list` here is the `messagesList`

File: src/main/java/com/easypost/model/ErrorDeserializer.java (hunk 4 above)
Comment (jchen293) [reply src/main/java/com/easypost/model/ErrorDeserializer.java:66]: Nate is correct, everything is being handled in the `traverseJsonElement` function recursively.

File: src/main/java/com/easypost/model/ErrorDeserializer.java (hunk 3 above)
Comment (Justintime50) [reply src/main/java/com/easypost/model/ErrorDeserializer.java:23]: mmm re-reading this I just misunderstood it

**Review by nwithan8 — APPROVED** [review]
